### PR TITLE
Implement User wallets (Uphold) on Android. (#8511)

### PR DESCRIPTION
**Pull request by gdregalo**
Resolves https://github.com/brave/brave-browser/issues/8511

## Submitter Checklist:

- [x] Submitted a [ticket](https://github.com/brave/brave-browser/issues) for my issue if one did not already exist.
- [x] Used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message.
- [ ] Added/updated tests for this change (for new code or code which already has tests).
- Verified that these changes build without errors on
  - [x] Android
  - [ ] iOS
  - [ ] Linux
  - [ ] macOS
  - [ ] Windows
- Verified that these changes pass automated tests (unit, browser, security tests) on
  - [ ] iOS
  - [ ] Linux
  - [ ] macOS
  - [ ] Windows
- [ ] Verified that all lint errors/warnings are resolved (`npm run lint`)
- [ ] Ran `git rebase master` (if needed).
- [ ] Ran `git rebase -i` to squash commits (if needed).
- [x] Tagged reviewers and labelled the pull request as needed.
- [ ] Request a security/privacy review as needed.
- [x] Add appropriate QA labels (QA/Yes or QA/No) to include the closed issue in milestone
- [ ] Public documentation has been updated as necessary. For instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protection-Mode
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-compatibility-issues-with-tracking-protection
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide

## Reviewer Checklist:

- [ ] New files have MPL-2.0 license header.
- [ ] Request a security/privacy review as needed.
- [ ] Adequate test coverage exists to prevent regressions 
- [ ] Verify test plan is specified in PR before merging to source

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on.
- [ ] All relevant documentation has been updated.

### CODE REVIEW

**Comment by gdregalo** [comment]
rebased on 46fc3e0b81595c7d38f891e842df160a01daa7ef

**Comment by gdregalo** [comment]
Ready for review!
I will re-base once Android build is fixed.

File: components/brave_rewards/browser/external_wallet.cc [hunk 1]
```diff
@@ -32,7 +32,9 @@ namespace brave_rewards {
     base::Value dict(base::Value::Type::DICTIONARY);
     dict.SetKey("token", base::Value(token));
     dict.SetKey("address", base::Value(address));
-    dict.SetKey("status", base::Value((int32_t)(status)));  //enum class WalletStatus : int32_t
+
+    // enum class WalletStatus : int32_t
+    dict.SetKey("status", base::Value((int32_t)(status)));
```
Comment (NejcZdovc) [inline components/brave_rewards/browser/external_wallet.cc:37]: please use static_cast

File: components/brave_rewards/browser/external_wallet.cc [hunk 2]
```diff
@@ -25,4 +27,21 @@ namespace brave_rewards {
...
   }
 
+  std::string ExternalWallet::toJson() {
+    std::string json_wallet;
+    base::Value dict(base::Value::Type::DICTIONARY);
+    dict.SetKey("token", base::Value(token));
+    dict.SetKey("address", base::Value(address));
+
+    // enum class WalletStatus : int32_t
+    dict.SetKey("status", base::Value((int32_t)(status)));
```
Comment (NejcZdovc) [inline components/brave_rewards/browser/external_wallet.cc:37]: please use static_cast

File: components/brave_rewards/browser/external_wallet.cc [hunk 3]
```diff
@@ -25,4 +27,21 @@ namespace brave_rewards {
     account_url = properties.account_url;
   }
 
+  std::string ExternalWallet::toJson() {
+    std::string json_wallet;
+    base::Value dict(base::Value::Type::DICTIONARY);
+    dict.SetKey("token", base::Value(token));
```
Comment (NejcZdovc) [inline components/brave_rewards/browser/external_wallet.cc:33]: `dict.SetKey("token", base::Value(token));` -> `dict.SetStringKey("token", token);`
please change ti to for all sets in this function

**Comment by gdregalo** [comment]
CI failed on unrelated issue:
https://bravesoftware.slack.com/archives/CA5FPHWLF/p1586829510286700?thread_ts=1586829436.286600&cid=CA5FPHWLF

File: components/brave_rewards/browser/external_wallet.cc (hunk 2 above)
Comment (gdregalo) [reply components/brave_rewards/browser/external_wallet.cc]: fixed

File: components/brave_rewards/browser/external_wallet.cc (hunk 3 above)
Comment (gdregalo) [reply components/brave_rewards/browser/external_wallet.cc]: fixed

**Comment by gdregalo** [comment]
All green! Please review.

File: android/java/org/chromium/chrome/browser/BraveRewardsExternalWallet.java [hunk 4]
```diff
@@ -0,0 +1,152 @@
+/** Copyright (c) 2019 The Brave Authors. All rights reserved.
```
Comment (SergeyZhukovsky) [inline android/java/org/chromium/chrome/browser/BraveRewardsExternalWallet.java:1]: 2020

File: android/java/org/chromium/chrome/browser/BraveRewardsExternalWallet.java [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+
+    public String mAccount_url;
+    public String mAdd_url;
+    public String mAddress;
+    @WalletStatus public int mStatus;
+    public String mToken;
+    public String mType;
+    public String mUser_name;
+    public String mVerify_url;
+    public String mWithdraw_url;
```
Comment (SergeyZhukovsky) [inline android/java/org/chromium/chrome/browser/BraveRewardsExternalWallet.java:151]: Should be probably moved on top of the class members following Java standard

File: patches/chrome-android-java-src-org-chromium-chrome-browser-externalnav-ExternalNavigationHandler.java.patch [hunk 6]
```diff
@@ -0,0 +1,13 @@
+diff --git a/chrome/android/java/src/org/chromium/chrome/browser/externalnav/ExternalNavigationHandler.java b/chrome/android/java/src/org/chromium/chrome/browser/externalnav/ExternalNavigationHandler.java
+index 929f86899690e5baa367af74186ccfb8b29df8ae..ddfcdd09679c758ef8d91a5d0e99ec8bb7e3d416 100644
+--- a/chrome/android/java/src/org/chromium/chrome/browser/externalnav/ExternalNavigationHandler.java
++++ b/chrome/android/java/src/org/chromium/chrome/browser/externalnav/ExternalNavigationHandler.java
+@@ -983,7 +983,7 @@ public class ExternalNavigationHandler {
+      * @return {@link OverrideUrlLoadingResult} if the tab was clobbered, or we launched an
+      *         intent.
+      */
+-    private @OverrideUrlLoadingResult int clobberCurrentTabWithFallbackUrl(
```
Comment (samartnik) [inline patches/chrome-android-java-src-org-chromium-chrome-browser-externalnav-ExternalNavigationHandler.java.patch:9]: This should be done via asm. See sample here https://github.com/brave/brave-core/blob/2c20620b3da7e8b55d61ec0a2fd61322fe6ec13e/build/android/bytecode/java/org/brave/bytecode/BraveBookmarkModelClassAdapter.java#L26

File: android/java/org/chromium/chrome/browser/BraveRewardsUserWalletActivity.java [hunk 7]
```diff
@@ -0,0 +1,131 @@
+/** Copyright (c) 2019 The Brave Authors. All rights reserved.
```
Comment (SergeyZhukovsky) [inline android/java/org/chromium/chrome/browser/BraveRewardsUserWalletActivity.java:1]: 2020

File: android/java/org/chromium/chrome/browser/BraveRewardsVerifyWalletActivity.java [hunk 8]
```diff
@@ -0,0 +1,72 @@
+/** Copyright (c) 2019 The Brave Authors. All rights reserved.
```
Comment (SergeyZhukovsky) [inline android/java/org/chromium/chrome/browser/BraveRewardsVerifyWalletActivity.java:1]: 2020

File: android/java/org/chromium/chrome/browser/BraveUphold.java [hunk 9]
```diff
@@ -0,0 +1,142 @@
+/** Copyright (c) 2019 The Brave Authors. All rights reserved.
```
Comment (SergeyZhukovsky) [inline android/java/org/chromium/chrome/browser/BraveUphold.java:1]: 2020

File: android/java/org/chromium/chrome/browser/externalnav/BraveExternalNavigationHandler.java [hunk 10]
```diff
@@ -0,0 +1,41 @@
+/* Copyright (c) 2019 The Brave Authors. All rights reserved.
```
Comment (SergeyZhukovsky) [inline android/java/org/chromium/chrome/browser/externalnav/BraveExternalNavigationHandler.java:1]: 2020

File: android/java/res/drawable/rect_round_corners_12.xml [hunk 11]
```diff
@@ -0,0 +1,9 @@
+<?xml version="1.0" encoding="utf-8"?>
+<!-- Copyright 2019 The Brave Authors. All rights reserved.
```
Comment (SergeyZhukovsky) [inline android/java/res/drawable/rect_round_corners_12.xml:2]: 2020

File: android/java/res/drawable/rect_up_round_corners_12.xml [hunk 12]
```diff
@@ -0,0 +1,10 @@
+<?xml version="1.0" encoding="utf-8"?>
+<!-- Copyright 2019 The Brave Authors. All rights reserved.
```
Comment (SergeyZhukovsky) [inline android/java/res/drawable/rect_up_round_corners_12.xml:2]: 2020

File: android/java/res/drawable/wallet_disconnected_button.xml [hunk 13]
```diff
@@ -0,0 +1,21 @@
+<?xml version="1.0" encoding="utf-8"?>
```
Comment (SergeyZhukovsky) [inline android/java/res/drawable/wallet_disconnected_button.xml:1]: Copyright?

File: android/java/res/drawable/wallet_verify_button.xml [hunk 14]
```diff
@@ -0,0 +1,21 @@
+<?xml version="1.0" encoding="utf-8"?>
```
Comment (SergeyZhukovsky) [inline android/java/res/drawable/wallet_verify_button.xml:1]: Copyright?

File: android/java/res/layout-land/verify_wallet_activity.xml [hunk 15]
```diff
@@ -0,0 +1,173 @@
+<?xml version="1.0" encoding="utf-8"?>
```
Comment (SergeyZhukovsky) [inline android/java/res/layout-land/verify_wallet_activity.xml:1]: Copyright?

File: android/java/res/layout/brave_rewards_panel_header.xml [hunk 16]
```diff
@@ -0,0 +1,238 @@
+<?xml version="1.0" encoding="utf-8"?>
```
Comment (SergeyZhukovsky) [inline android/java/res/layout/brave_rewards_panel_header.xml:1]: Copyright?

File: android/java/res/layout/user_wallet_activity.xml [hunk 17]
```diff
@@ -0,0 +1,146 @@
+<?xml version="1.0" encoding="utf-8"?>
+<!--* Copyright (c) 2019 The Brave Authors. All rights reserved.
```
Comment (SergeyZhukovsky) [inline android/java/res/layout/user_wallet_activity.xml:2]: 2020

File: android/java/res/layout/verify_wallet_activity.xml [hunk 18]
```diff
@@ -0,0 +1,138 @@
+<?xml version="1.0" encoding="utf-8"?>
+<!--* Copyright (c) 2019 The Brave Authors. All rights reserved.
```
Comment (SergeyZhukovsky) [inline android/java/res/layout/verify_wallet_activity.xml:2]: 2020

File: android/java/res/layout/verify_wallet_activity.xml [hunk 19]
```diff
@@ -0,0 +1,138 @@
...
+        android:id="@+id/service_provider_txt"
+        android:layout_width="wrap_content"
+        android:layout_height="wrap_content"
+        android:layout_gravity="center"
+        android:gravity="center_vertical"
+        android:layout_marginTop="@dimen/verify_wallet_margin2"
+        android:drawableEnd="@drawable/uphold_green"
+        android:drawablePadding="6dp"
+        android:text="@string/verify_wallet_service_note"
+        android:textColor="#e6e3f4"
```
Comment (SergeyZhukovsky) [inline android/java/res/layout/verify_wallet_activity.xml:116]: let's move it to `brave_colors.xml`

File: android/java/res/layout/verify_wallet_activity.xml [hunk 20]
```diff
@@ -0,0 +1,138 @@
...
+        android:textSize="@dimen/verify_wallet_header4_font_size"/>
+
+    <TextView
+        android:layout_width="wrap_content"
+        android:layout_height="wrap_content"
+        android:layout_gravity="start"
+        android:layout_marginTop="@dimen/verify_wallet_margin3"
+        android:text="@string/verify_wallet_uphold_note1"
+        android:textAlignment="textStart"
+        android:textColor="#e6e3f4"
```
Comment (SergeyZhukovsky) [inline android/java/res/layout/verify_wallet_activity.xml:126]: let's move it to `brave_colors.xml`

File: android/java/res/layout/verify_wallet_activity.xml [hunk 21]
```diff
@@ -0,0 +1,138 @@
...
+        android:textSize="@dimen/verify_wallet_header4_font_size"/>
+
+    <TextView
+        android:layout_width="wrap_content"
+        android:layout_height="wrap_content"
+        android:layout_gravity="start"
+        android:layout_marginTop="@dimen/verify_wallet_margin3"
+        android:text="@string/verify_wallet_brave_note1"
+        android:textAlignment="textStart"
+        android:textColor="#e6e3f4"
```
Comment (SergeyZhukovsky) [inline android/java/res/layout/verify_wallet_activity.xml:136]: let's move it to `brave_colors.xml`

File: android/java/res/values-h380dp/dimens.xml [hunk 22]
```diff
@@ -0,0 +1,18 @@
+<?xml version="1.0" encoding="utf-8"?>
+<!-- Copyright (c) 2019 The Brave Authors. All rights reserved.
```
Comment (SergeyZhukovsky) [inline android/java/res/values-h380dp/dimens.xml:2]: 2020

File: android/java/res/values-h730dp/dimens.xml [hunk 23]
```diff
@@ -0,0 +1,18 @@
+<?xml version="1.0" encoding="utf-8"?>
+<!-- Copyright (c) 2019 The Brave Authors. All rights reserved.
```
Comment (SergeyZhukovsky) [inline android/java/res/values-h730dp/dimens.xml:2]: 2020

File: android/java/org/chromium/chrome/browser/BraveRewardsVerifyWalletActivity.java (hunk 8 above)
Comment (gdregalo) [reply android/java/org/chromium/chrome/browser/BraveRewardsVerifyWalletActivity.java]: fixed

File: android/java/org/chromium/chrome/browser/BraveUphold.java (hunk 9 above)
Comment (gdregalo) [reply android/java/org/chromium/chrome/browser/BraveUphold.java]: fixed

File: android/java/org/chromium/chrome/browser/externalnav/BraveExternalNavigationHandler.java (hunk 10 above)
Comment (gdregalo) [reply android/java/org/chromium/chrome/browser/externalnav/BraveExternalNavigationHandler.java]: fixed

File: android/java/res/drawable/rect_round_corners_12.xml (hunk 11 above)
Comment (gdregalo) [reply android/java/res/drawable/rect_round_corners_12.xml]: fixed

File: android/java/res/drawable/rect_up_round_corners_12.xml (hunk 12 above)
Comment (gdregalo) [reply android/java/res/drawable/rect_up_round_corners_12.xml]: fixed

File: android/java/org/chromium/chrome/browser/BraveRewardsUserWalletActivity.java (hunk 7 above)
Comment (gdregalo) [reply android/java/org/chromium/chrome/browser/BraveRewardsUserWalletActivity.java]: fixed

File: android/java/res/drawable/wallet_verify_button.xml (hunk 14 above)
Comment (gdregalo) [reply android/java/res/drawable/wallet_verify_button.xml:1]: fixed

File: android/java/res/drawable/wallet_disconnected_button.xml (hunk 13 above)
Comment (gdregalo) [reply android/java/res/drawable/wallet_disconnected_button.xml:1]: fixed

File: android/java/res/layout-land/verify_wallet_activity.xml (hunk 15 above)
Comment (gdregalo) [reply android/java/res/layout-land/verify_wallet_activity.xml:1]: fixed

File: android/java/res/layout/brave_rewards_panel_header.xml (hunk 16 above)
Comment (gdregalo) [reply android/java/res/layout/brave_rewards_panel_header.xml:1]: fixed

File: android/java/res/layout/user_wallet_activity.xml (hunk 17 above)
Comment (gdregalo) [reply android/java/res/layout/user_wallet_activity.xml]: fixed

File: android/java/res/layout/verify_wallet_activity.xml (hunk 18 above)
Comment (gdregalo) [reply android/java/res/layout/verify_wallet_activity.xml]: fixed

File: android/java/res/values-h380dp/dimens.xml (hunk 22 above)
Comment (gdregalo) [reply android/java/res/values-h380dp/dimens.xml]: fixed

File: android/java/res/values-h730dp/dimens.xml (hunk 23 above)
Comment (gdregalo) [reply android/java/res/values-h730dp/dimens.xml]: fixed

File: android/java/res/layout/verify_wallet_activity.xml (hunk 19 above)
Comment (gdregalo) [reply android/java/res/layout/verify_wallet_activity.xml]: fixed

File: android/java/res/layout/verify_wallet_activity.xml (hunk 20 above)
Comment (gdregalo) [reply android/java/res/layout/verify_wallet_activity.xml]: fixed

File: android/java/res/layout/verify_wallet_activity.xml (hunk 21 above)
Comment (gdregalo) [reply android/java/res/layout/verify_wallet_activity.xml]: fixed

File: patches/chrome-android-java-src-org-chromium-chrome-browser-externalnav-ExternalNavigationHandler.java.patch (hunk 6 above)
Comment (gdregalo) [reply patches/chrome-android-java-src-org-chromium-chrome-browser-externalnav-ExternalNavigationHandler.java.patch]: fixed

**Comment by gdregalo** [comment]
It's all green again. Please review.

**Comment by gdregalo** [comment]
@NejcZdovc  
Connected state is implemented and reflected on UI in accordance to state diagram at
https://drive.google.com/drive/folders/1AwF9rC7L8Ru9oz4i0gp0ui1GeHWjXiQ1.
There is nothing about Site banner there. The rewards html ui is not updated as discussed.

**Comment by NejcZdovc** [comment]
@gdregalo diagram doesn't have publisher flows it only has uphold flows

**Comment by gdregalo** [comment]
@NejcZdovc I guess I should have been provided the document describing the publisher flow in the beginning of the task instead of  when I want to merge it.

**Comment by NejcZdovc** [comment]
@gdregalo documents were provided to you by @kylehickinson

**Comment by gdregalo** [comment]
@NejcZdovc 
1. Added publisher's CONNECTED state note for both panel and site banner.
2. Used white Uphold icon.
3. Known and hard to reproduce. It will be reviewed in a separate issue.
4. The Uphold page has to be redirected somewhere. We can not close the tab because it might be the last one and it will cause the app to close too. Leaving it in indeterminate state is not an option either. The solution can be to open the rewards settings page where the latest wallet status is reflected. But it doesn't have it at the moment (WIP) . As temporary solution: opening both the panel and rewards page.

**Comment by gdregalo** [comment]
It's green again. Please review.

**Comment by gdregalo** [comment]
@bridiver 
Created issue: https://github.com/brave/brave-browser/issues/9395
"Rewrite TabbedModeTabDelegateFactory.java.patch using ASM"

**Pull request merged**